### DISCUSSION TITLE
How to use types and enums generated by prisma in Next.js frontend.

### DISCUSSION THREAD

**Discussion by toiroakr** [prisma/prisma#10137] [Q&A] [answered]
## Problem
In Next.js frontend, I cannot import models with error. (Next.js v12.0.1 / Prisma v3.2.1)
```
error - ../model/node_modules/@prisma/client/runtime/index.js:29065:0
Module not found: Can't resolve '_http_common'

Import trace for requested module:
../model/node_modules/.prisma/client/index.js
../model/node_modules/@prisma/client/index.js
../model/user.ts
./pages/index.tsx
```

#### directories
```
root
 ├ backend
 ├ frontend
 │  └ pages/index.ts
 └ model
    ├ prisma/schema.prisma
    └ user.ts
```
#### frontend/tsconfig.ts
```
{
  "compilerOptions": {
    "paths": { "@model": ["../model"] },
...
  "references": [{ "path": "../model" }]
}
```
#### schema.prisma
```prisma
model User {
  id        String       @id @default(cuid())
  createdAt DateTime     @default(now())
}

enum Role {
  USER
  ADMIN
}
```
#### user.ts
```ts
export { User, Role } from "@prisma/client";
```

#### frontend/pages/index.tsx
```ts
import { Role } from "@noxas/model";

export default function Page() {
  return <div>{Role.USER}</div>
}
```

## Question 
How to use types and enums in Next.js frontend?
I want to use just types and enum, not PrismaClient. Is there a way to import just types and enums?
`export type` is not fine for me because enums are const.

**Comment by austincrim** [comment]
Hi @toiroakr 👋,

You should be able to import any generated Prisma type in your frontend without an issue. Would you be able to share a minimal reproduction repo so I can dig into this?

Thanks!

**Reply by toiroakr** [reply]
Hi @austincrim 👋,
Thank you for replying.

Here is minimal reproduction repo for this issue.
https://github.com/toiroakr/prisma-issue-sample 

When I was creating it, I thought that webpack setting for ts-loader (or its combination with ts-config) might be wrong. 🤔
![image](https://user-images.githubusercontent.com/3090700/141679816-903c2857-5e29-4a95-ba5d-697f5939c09e.png)

**Reply by austincrim** [reply]
Hey @toiroakr, thanks so much for the reproduction!

I was able to get things working by making a couple changes in `_app.tsx`:

```tsx
import { Fragment } from 'react'
import type { AppProps } from 'next/app'
import type { Role } from '@model'

export default function App({ Component, pageProps }: AppProps) {
  const role: Role = 'USER'
  return (
    <Fragment>
      <div>{role}</div>
    </Fragment>
  )
}
```

I think the root of the issue was trying to use `Role` as a value in the JSX instead of just using it as a type, which was your goal I believe.

Let me know if you have any questions!

**Reply by toiroakr** [reply]
Thanks @austincrim !

Role defined in: `models/node_modules/.prisma/client/index.d.ts` below;
```typescript
export const Role: {
  USER: 'USER',
  ADMIN: 'ADMIN'
};

export type Role = (typeof Role)[keyof typeof Role]
```

I want to use Role as const value but not type.(in this case)
for example:
```
<ul>
  {Object.values(Role).map(role => <li>{role}</li>)}
</ul>

// or
const function canEdit(role: Role) {
  return role === Role.ADMIN
}
```

**Reply by toiroakr** [reply]
I currently work around this as follows. ( https://github.com/toiroakr/prisma-issue-sample/commit/3c365242649748d3f7ce49b8ecd24a9ce8bd322a )

`generator.js`
```javascript
const helper = require('@prisma/generator-helper');
const fs = require('fs');
const path = require('path');

helper.generatorHandler({
  onManifest() {
    return {
      defaultOutput: './types/_prisma',
      prettyName: 'Prisma Type Only Generator',
    };
  },
  async onGenerate(options) {
    const output = options.generator.output?.value;

    if (output) {
      try {
        // content are "Model *" & "Enums" extracted from "node_modules/.prisma/client/index.d.ts".
        let content = fs.readFileSync(path.resolve('node_modules/.prisma/client/index.d.ts')).toString();
        content = content.replace(/\/\*\*\n \* Client([\s\S]+?)\/\*\*/g, '/**')
        content = content.replace(/\/\*\*\n \* ## {2}Prisma Client[\s\S]+/g, '');

        // replace "export const Role:" to "export const Role ="
        content = content.replace(/(export const \w+):/g, '$1 =').trim();

        fs.mkdirSync(output, { recursive: true });
        fs.writeFileSync(path.join(output, 'index.ts'), content);
      } catch (error) {
        console.error(error);
        throw error;
      }
    } else {
      throw new Error('No output specified for Prisma Type Only Generator');
    }
  },
});
```

`schema.prisma`
```
...
generator model_type {
  provider = "node ./generator.js"
  output   = "../models/_prisma"
}
...
```

`model/models/index.ts`
```typescript
export { Role } from "./_prisma";
```

![image](https://user-images.githubusercontent.com/3090700/141884493-078b473e-70d4-4042-be21-9b9210675a27.png)

**Reply by toiroakr** [reply]
@austincrim 
The workaround is dirty, is there anything better else?

As for me, It would be nice if the output of prisma-client-js is divided into "models & enums" and "Prisma & PrismaClient", then can import separatedly.

**Reply by toiroakr** [reply]
I made another workaround with prisma@4. (It still dirty...)
https://github.com/toiroakr/prisma-issue-sample/pull/1

**Reply by toiroakr** [reply]
I made solution with prisma@4. (It's better way.)
https://github.com/toiroakr/prisma-issue-sample/pull/2

**Comment by basememara** [comment]
This is still an issue with Prisma v5. Why can't we use Prisma types and enums on the client-side? Is this something that can be overcome?

I'm purely using the Prisma definition file on my frontend.. it doesn't even have a dependency on Prisma so I'm baffled how this can be an issue using just the `node_modules/.prisma/client/index.d.ts` file.

**Comment by gruckion** [comment]
Please address this issue prisma.

**Comment by electrovir** [comment, marked as answer]
I wrote my own generator to solve this for myself: https://www.npmjs.com/package/prisma-frontend

**Reply by ludralph** [reply]
Thanks for creating this. It would be really helpful to the community.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
